### ISSUE TITLE
Shouldn't `Carbon::withTestNow()` restore original `$testNow` after invoking `$callback()`?

### ISSUE THREAD

**Issue by mpyw** [briannesbitt/Carbon#2882]
Hello,

I encountered an issue with the following code:

```php
Carbon::setTestNow('2023-01-01 00:00:00');
Carbon::withTestNow('2023-01-02 00:00:00', fn () => doSomething());
echo Carbon::now()->format('Y-m-d H:i:s');
```

Carbon version:  2.71.0

PHP version: 8.2

I expected to get:

```
2023-01-01 00:00:00
```

But I actually get:

```
2023-11-17 18:09:00 (Real Time)
```

Thanks!

**Comment by kylekatarnls** [comment]
Hello. I get the idea, as per method documentation, it's clearing the test instance after callback is executed, this has been the behavior since this method is there (more than 3 years), so changing it now is likely to break test suite for a lot of people.

However you can create your own alternative macro like this:

```php
// Run this in the tests bootstrap file

Carbon::macro('runWithTestNow', static function ($testNow, $callback) {
    $previousNow = Carbon::getTestNow();
    Carbon::setTestNow($testNow);

    try {
        $result = $callback();
    } finally {
        Carbon::setTestNow($previousNow);
    }

    return $result;
});
```

Then you can use in your tests `Carbon::runWithTestNow('2023-01-02 00:00:00', fn () => doSomething())` which has the behavior you want.

**Comment by mpyw** [comment]
@kylekatarnls Is it okay for branch 3.x?

**Comment by kylekatarnls** [comment]
Actually I would consider deleting this method in 3.x.

Reason is for that is mostly interoperability.

We implemented [carbonite](https://github.com/kylekatarnls/carbonite) for advanced testing tools, then `withTestNow()` was added, then Laravel also implemented yet another way to mock time in test later with the same features which is using the Carbon's one under the hood. And Symfony has a different one based on clock. But if you have Symfony+Carbon, they won't be in sync.

Carbon also has different testNow for mutable and immutable class. Some users rely on that but in real life you would never have them at different time, so it shouldn't be like that neither.

Then PSR introduced [PSR-20: Clock](https://www.php-fig.org/psr/psr-20/) which aims to standardize it. And I think this is the direction to take.

Having as less test tooling as possible in the Carbon library itself, instead having just a way to plug to a clock (first step was done by implementing the psr-clock). Then testing tooling could interact with the clock so each component stick to 1 responsibility and you would less likely end with time inconsistency in tests.
